src/init/index.ts

```diff
@@ -1,43 +1,50 @@
-import * as _ from "lodash";
+import { capitalize } from "lodash";
 import * as clc from "cli-color";
+
+import { FirebaseError } from "../error";
 import { logger } from "../logger";
-import * as _features from "./features";
-import * as utils from "../utils";
+import * as features from "./features";
 
-export interface Indexable<T = any> {
-  [key: string]: T;
-}
-export interface RCFile {
-  projects: Indexable;
-}
 export interface Setup {
-  config: Indexable;
-  rcfile: RCFile;
+  config: Record<string, any>;
+  rcfile: {
+    projects: Record<string, any>;
+  };
   features?: string[];
   featureArg?: boolean;
-  project?: Indexable;
+  project?: Record<string, any>;
   projectId?: string;
   projectLocation?: string;
 }
 
-// TODO: Convert features/index.js to TypeScript so it exports
-// as an indexable type instead of doing this cast.
-const features = _features as Indexable;
+const featureFns = new Map<string, any>([
+  ["account", features.account],
+  ["database", features.database],
+  ["firestore", features.firestore],
+  ["functions", features.functions],
+  ["hosting", features.hosting],
+  ["storage", features.storage],
+  ["emulators", features.emulators],
+  ["project", features.project], // always runs, sets up .firebaserc
+  ["remoteconfig", features.remoteconfig],
+  ["hosting:github", features.hostingGithub],
+]);
 
 export async function init(setup: Setup, config: any, options: any): Promise<any> {
   const nextFeature = setup.features ? setup.features.shift() : undefined;
   if (nextFeature) {
-    if (!features[nextFeature]) {
-      return utils.reject(
-        clc.bold(nextFeature) +
-          " is not a valid feature. Must be one of " +
-          _.without(_.keys(features), "project").join(", ")
+    if (!featureFns.has(nextFeature)) {
+      const availableFeatures = Object.keys(features)
+        .filter((f) => f !== "project")
+        .join(", ");
+      throw new FirebaseError(
+        `${clc.bold(nextFeature)} is not a valid feature. Must be one of ${availableFeatures}`
       );
     }
 
-    logger.info(clc.bold("\n" + clc.white("=== ") + _.capitalize(nextFeature) + " Setup"));
+    logger.info(clc.bold(`\n${clc.white("===")} ${capitalize(nextFeature)} Setup`));
 
-    await Promise.resolve(features[nextFeature](setup, config, options));
+    await Promise.resolve(featureFns.get(nextFeature)(setup, config, options));
```

Comment:
q: what does this extra `Promise.resolve` do 🤔

Reply:
It ensures that whatever is being called _is_ a promise. If it's a promise, wrapping it one other promise is just a bit of overhead; if it's a sync function, it means that we await on it anyway. It's a safety check, more than anything. If any of the setup functions don't return a promise, this protected against that.

But now that it's all in typescript, and we _know_ the types of the functions, I can actually drop this extra promise! yay!